protocol-units/da-sequencer/node/src/storage/mod.rs

```diff
@@ -1,36 +1,124 @@
+}
 
 impl Storage {
+	pub fn try_new(path: &str) -> Result<Self, DaSequencerError> {
+		let mut options = Options::default();
+
+		options.create_if_missing(true);
+		options.create_missing_column_families(true);
+
+		let pending_transactions_cf =
+			ColumnFamilyDescriptor::new(cf::PENDING_TRANSACTIONS, Options::default());
+		let blocks_cf = ColumnFamilyDescriptor::new(cf::BLOCKS, Options::default());
+		let blocks_by_digest_cf =
+			ColumnFamilyDescriptor::new(cf::BLOCKS_BY_DIGEST, Options::default());
+
+		let db = DB::open_cf_descriptors(
+			&options,
+			path,
+			[pending_transactions_cf, blocks_cf, blocks_by_digest_cf],
+		)
+		.map_err(|e| DaSequencerError::Generic(e.to_string()))?;
+
+		Ok(Storage { db: Arc::new(db) })
+	}
+
 	/// Save all batch's Tx in the pending Tx table. The batch's Tx has been verified and validated.
 	pub fn write_batch(
 		&self,
-		batch: DaBatch<FullnodeTx>,
+		batch: DaBatch<FullNodeTx>,
 	) -> std::result::Result<(), DaSequencerError> {
-		todo!();
+		let cf = self.db.cf_handle(cf::PENDING_TRANSACTIONS).ok_or_else(|| {
+			DaSequencerError::Generic("Missing column family: pending_transactions".into())
+		})?;
+		let tx = batch.data;
+		let key = tx.id.0;
+		let value = bincode::serialize(&tx)
+			.map_err(|e| DaSequencerError::Generic(format!("Serialization error: {}", e)))?;
+
+		let mut write_batch = WriteBatch::default();
+		write_batch.put_cf(&cf, key, value);
+
+		self.db
+			.write(write_batch)
+			.map_err(|e| DaSequencerError::Generic(format!("DB write error: {}", e)))?;
+
+		Ok(())
 	}
 
 	/// Return, if exists, the Block at the given height.
 	pub fn get_block_at_height(
 		&self,
 		height: BlockHeight,
 	) -> std::result::Result<Option<SequencerBlock>, DaSequencerError> {
-		todo!();
+		let cf = self
+			.db
+			.cf_handle(cf::BLOCKS)
+			.ok_or_else(|| DaSequencerError::Generic("Missing column family: blocks".into()))?;
+
+		let key: [u8; 8] = height.0.to_be_bytes();
+
+		match self.db.get_cf(&cf, key).map_err(|e| DaSequencerError::Generic(e.to_string()))? {
+			Some(bytes) => {
+				let block: SequencerBlock = bincode::deserialize(&bytes).map_err(|e| {
+					DaSequencerError::Generic(format!("Deserialization error: {}", e))
+				})?;
+				Ok(Some(block))
+			}
+			None => Ok(None),
+		}
 	}
 
 	/// Return, if exists, the Block with specified sequencer id.
 	pub fn get_block_with_digest(
 		&self,
 		id: SequencerBlockDigest,
 	) -> std::result::Result<Option<SequencerBlock>, DaSequencerError> {
-		todo!();
+		// Step 1: Get the height from the digest → height mapping
+		let cf = self.db.cf_handle(cf::BLOCKS_BY_DIGEST).ok_or_else(|| {
+			DaSequencerError::Generic("Missing column family: blocks_by_digest".into())
+		})?;
+
+		let key = id.0;
+
+		let height_bytes =
+			match self.db.get_cf(&cf, key).map_err(|e| DaSequencerError::Generic(e.to_string()))? {
+				Some(bytes) => bytes,
+				None => return Ok(None),
+			};
+
+		if height_bytes.len() != 8 {
+			return Err(DaSequencerError::Generic(
```

Comment:
Could be a storage error or more specifically, storage format error (where also deserialization failures would fit, since it's the same problem of the database entry format being seemingly wrong)